DOC: add stateful computation doc (#20732)
jax-ml/jax: docs/tutorials/stateful-computations.md

---
jupytext:
  formats: md:myst
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.16.1
kernelspec:
  display_name: Python 3
  language: python
  name: python3
---

# Stateful Computations

JAX transformations like {func}`~jax.jit`, {func}`~jax.vmap`, {func}`~jax.grad`, require the functions
they wrap to be pure: that is, functions whose outputs depend *solely* on the inputs, and which have

> no side-effects such as updating of global state.

Reviewer: ```suggestion
no side effects such as updating of global state.
```

This constraint can pose some challenges in the context of machine learning, where state may exist in
many forms. For example:

* model parameters,
* optimizer state, and
* stateful layers, such as [BatchNorm](https://en.wikipedia.org/wiki/Batch_normalization).

This section offers some advice of how to properly handle state in a JAX program.

## A simple example: Counter

Let's start by looking at a simple stateful program: a counter.

```{code-cell}
import jax
import jax.numpy as jnp

class Counter:
  """A simple counter."""

  def __init__(self):
    self.n = 0

  def count(self) -> int:
    """Increments the counter and returns the new value."""
    self.n += 1
    return self.n

  def reset(self):
    """Resets the counter to zero."""
    self.n = 0

counter = Counter()

for _ in range(3):
  print(counter.count())
```

The counter's `n` attribute maintains the counter's _state_ between successive calls of `count`. It is modified as a side effect of calling `count`.

Let's say we want to count fast, so we JIT-compile the `count` method.

> (In this example, this wouldn't actually help speed anyway, for many reasons, but treat this as a toy model of wanting to JIT-compile the update of model parameters, where {func}`~jax.jit` makes an enormous difference).

Reviewer: ```suggestion
(In this example, this wouldn't actually help speed anyway, for many reasons, but treat this as a toy model of JIT-compiling the update of model parameters, where {func}`~jax.jit` makes an enormous difference).
```

```{code-cell}
counter.reset()
fast_count = jax.jit(counter.count)

for _ in range(3):
  print(fast_count())
```

Oh no! Our counter isn't working. This is because the line
```
self.n += 1
```

> in `count` is a side-effect: it modifies the input counter in-place, and so this function is not supported by `jit`.

Reviewer: ```suggestion
in `count` involves a side effect: it modifies the input counter in-place, and so this function is not supported by `jit`.
```

> Such side-effects are executed only once when the function is first traced, and subsequent calls will not repeat the side-effect.

Reviewer: ```suggestion
Such side effects are executed only once when the function is first traced, and subsequent calls will not repeat the side effect.
```
